### PR TITLE
kv: pass explicit Now timestamp on BatchRequest, remove timestamp downcasting

### PR DESCRIPTION
**Pull request by nvb**
This commit adds an explicit `ClockTimestamp` field called `Now` to the `BatchRequest` header, which mirrors the `Now` field on the `BatchResponse` header.

In doing so, it removes the last instance where we downcasted a `Timestamp` to a `ClockTimestamp` using the `TryToClockTimestamp` method. With this change, MVCC ("operation") timestamps never flow back into HLC clocks as clock signals. This was enabled by #80706 and sets the groundwork to remove synthetic timestamps in v23.1 — the role they played in dynamic typing of clock timestamps is now entirely fulfilled by statically typed ClockTimestamp channels.

This is an important step in separating out the MVCC timestamp domain from the clock timestamp domain and clarifying the roles of the two layers. In turn, this layering opens the door for CockroachDB to start thinking about dynamic clock synchronization error bounds.

### CODE REVIEW

**Review by tbg — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85764#-:-N97QUKn2GJ-5BhOcVLg:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Comment by nvanbenschoten** [comment]
TFTR!

Was this implicitly an approval for https://github.com/cockroachdb/cockroach/pull/76233 as well, or does that still need another look?

File: pkg/util/hlc/timestamp.go [hunk 1]
```diff
@@ -386,16 +386,19 @@ func (LegacyTimestamp) SafeValue() {}
...
 
-// TryToClockTimestamp attempts to downcast a Timestamp into a ClockTimestamp.
-// Returns the result and a boolean indicating whether the cast succeeded.
-func (t Timestamp) TryToClockTimestamp() (ClockTimestamp, bool) {
+// DeprecatedTryToClockTimestamp attempts to downcast a Timestamp into a
+// ClockTimestamp. Returns the result and a boolean indicating whether the cast
+// succeeded.
+// TODO(nvanbenschoten): remove this in v23.1 when we remove the synthetic
+// timestamp bit.
+func (t Timestamp) DeprecatedTryToClockTimestamp() (ClockTimestamp, bool) {
```
Comment (tbg) [inline pkg/util/hlc/timestamp.go:401]: 🎉

**Review by tbg — APPROVED** [review]

**Comment by nvanbenschoten** [comment]
TFTR!

bors r+

**Comment by nvanbenschoten** [comment]
bors r-

This seems to be causing issues in bors. CI was all green until the last rebase, so something might be up.

**Comment by nvanbenschoten** [comment]
I think it might have skewed with https://github.com/cockroachdb/cockroach/pull/83213. The benchmark (`BenchmarkMVCCGCWithForegroundTraffic`) added in that PR now seems to be failing.

**Pull request merged**